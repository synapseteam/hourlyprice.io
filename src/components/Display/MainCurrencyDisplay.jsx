import React from "react";
import PropTypes from "prop-types";

import Loader from "components/Loader";

export default function MainCurrencyDisplay({ sum, currency, isLoading }) {
  return (
    <>
      <h2 className="display-title">Total price</h2>
      <div className="display-sum-container">
        {isLoading ? (
          <Loader />
        ) : (
          <p className="display-sum">{`${sum}${currency}`}</p>
        )}
      </div>
    </>
  );
}

MainCurrencyDisplay.propTypes = {
  sum: PropTypes.string,
  currency: PropTypes.string,
  isLoading: PropTypes.bool,
};

MainCurrencyDisplay.defaultProps = {
  sum: 0.0,
  currency: "USD",
  isLoading: false,
};
