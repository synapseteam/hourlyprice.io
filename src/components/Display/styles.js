import { css } from "@emotion/react";
import { getStyle } from "utils/generic";

import {
  lightPurple,
  brightGrey,
} from "components/UI/sharedStylesEmotion/colors.js";

const mixinShadow = `
-webkit-box-shadow: 0px 18px 23px -3px rgba(0, 0, 0, 0.51);
-moz-box-shadow: 0px 18px 23px -3px rgba(0, 0, 0, 0.51);
box-shadow: 0px 18px 23px -3px rgba(0, 0, 0, 0.51);
`;

const mixinNoShadow = `
-webkit-box-shadow: none;
-moz-box-shadow: none;
box-shadow: none;
`;

export const styles = {
  display: css`
    background-color: ${lightPurple};
    border-radius: 0.4rem;
    position: relative;
    width: 100%;
    padding: 3rem 2rem;
    display: flex;
    flex-direction: column;
    align-items: center;
    margin-top: 1.2rem;
    ${mixinShadow};

    @media (min-width: 720px) and (max-width: 919px) {
      width: 47%;
      padding: 3rem 1rem;
    }

    @media (min-width: 920px) {
      width: 47%;
    }
  `,
  displayLight: css`
    background-color: ${brightGrey};
    ${mixinNoShadow}
  `,
  getStyle,
};
