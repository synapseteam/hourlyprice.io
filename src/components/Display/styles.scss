$light-purple: #31335a;
$blue: #5599ec;

@mixin box-shadow {
  -webkit-box-shadow: 0px 18px 23px -3px rgba(0, 0, 0, 0.51);
  -moz-box-shadow: 0px 18px 23px -3px rgba(0, 0, 0, 0.51);
  box-shadow: 0px 18px 23px -3px rgba(0, 0, 0, 0.51);
}

.display {
  background-color: $light-purple;
  border-radius: 0.4rem;
  width: 100%;
  padding: 3rem 2rem;
  margin-top: 2rem;

  display: flex;
  flex-direction: column;
  align-items: center;

  @include box-shadow;

  .display-title {
    margin: 0;
  }

  .display-sum-container {
    border-bottom: 1px solid lighten($light-purple, 15);
    min-height: 6.8rem;

    .display-sum {
      color: $blue;
      font-weight: 700;
      line-height: 0.7;
      padding: 2.4rem 1rem;
      margin: 0;
      font-size: 3rem;
    }
  }

  .display-currencies-container {
    margin-top: 1.6rem;
    min-height: 3rem;
    display: flex;
    width: 100%;
    justify-content: space-around;

    .display-sep-currency {
      display: flex;
      flex-direction: column;
      align-items: center;

      .display-sep-currency-type {
        color: $blue;
        font-size: 0.8rem;
      }

      .display-sep-currency-sum {
        margin-top: 0.4rem;
        font-weight: 700;
      }
    }
  }
}
