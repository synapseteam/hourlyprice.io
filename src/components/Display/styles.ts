import { css, Theme } from "@emotion/react";

export const styles = {
  display: (theme: Theme) => css`
    background-color: ${theme.primary};
    border-radius: 0.4rem;
    position: relative;
    width: 47%;
    padding: 3rem 2rem;
    display: flex;
    flex-direction: column;
    align-items: center;
    margin-top: 1.2rem;
    box-shadow: ${theme.shadow};
    text-align: center;
    @media (max-width: 420px) {
      padding: 3rem 0.5rem;
    }
    @media (max-width: 720px) {
      width: 100%;
    }
  `,
};
