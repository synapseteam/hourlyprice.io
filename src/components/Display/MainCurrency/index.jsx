/** @jsxImportSource @emotion/react */
import React from "react";
import PropTypes from "prop-types";

import { useCustomTranslation } from "i18n";
import Loader from "components/UI/Loader";

import { styles } from "./styles";

export default function MainCurrencyDisplay({
  sum,
  currency,
  isLoading,
  darkMode,
}) {
  const [t] = useCustomTranslation();
  return (
    <>
      <h2 css={() => styles.getStyle(darkMode, "title")}>{t("totalPrice")}</h2>
      <div css={() => styles.getStyle(darkMode, "mainContainer")}>
        {isLoading ? (
          <Loader />
        ) : (
          <p
            css={() => styles.getStyle(darkMode, "sum")}
          >{`${sum}${currency}`}</p>
        )}
      </div>
    </>
  );
}

MainCurrencyDisplay.propTypes = {
  sum: PropTypes.string,
  currency: PropTypes.string,
  isLoading: PropTypes.bool,
  darkMode: PropTypes.bool,
};

MainCurrencyDisplay.defaultProps = {
  sum: 0.0,
  currency: "USD",
  isLoading: false,
  darkMode: true,
};
