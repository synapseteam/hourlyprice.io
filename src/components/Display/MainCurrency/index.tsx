/** @jsxImportSource @emotion/react */

import { useCustomTranslation } from "i18n";
import SkeletonLoader from "components/UI/SkeletonLoader";

import { styles } from "./styles";

const initSum = 0.0;

interface IProps {
  sum: number;
  currency: string;
  isLoading: boolean;
}

const MainCurrencyDisplay: React.FC<IProps> = ({
  sum = initSum,
  currency = "USD",
  isLoading = false,
}): JSX.Element => {
  const [t] = useCustomTranslation();
  return (
    <>
      <h2 css={styles.title}>{t("totalPrice")}</h2>
      <div css={styles.mainContainer}>
        {isLoading ? (
          <SkeletonLoader size="l" />
        ) : (
          <p css={styles.sum}>{`${sum}${currency}`}</p>
        )}
      </div>
    </>
  );
};

export default MainCurrencyDisplay;
