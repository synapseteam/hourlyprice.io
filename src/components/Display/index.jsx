/** @jsxImportSource @emotion/react */
import React from "react";
import { useSelector } from "react-redux";

import SubCurrenciesDisplay from "components/Display/SubCurrencies";
import MainCurrencyDisplay from "components/Display/MainCurrency/";
import SubCurrenciesRatesDisplay from "components/Display/SubCurrenciesRates";
import { useAppThemeContext } from "context/AppContext";
import { useCustomTranslation } from "i18n";

import { styles } from "./styles";

export default function Display() {
  const { price, time, currency } = useSelector((state) => state.main.fields);

  const allCurrencies = useSelector((state) => state.rates.allCurrencies);

  const isLoading = useSelector((state) => state.main.isLoading);

  const isRequestError = useSelector((state) => state.main.ratesRequestErr);

  const [context] = useAppThemeContext();

  const [t] = useCustomTranslation();

  const mainCurrencyData = allCurrencies.find((el) => el.name === currency) || {
    name: "",
    price: 0,
    symbol: "",
  };

  function formatSum(num) {
    return new Intl.NumberFormat("de-DE")
      .format(num.toFixed(2))
      .replace(".", " ");
  }

  function generateSubCurrenciesArr(arr, basicRate) {
    const subCurrencyInitial = arr.filter((el) => el.name !== currency);
    if (!currency) {
      return subCurrencyInitial.map((el) => {
        return {
          ...el,
          value: 0,
        };
      });
    }

    return subCurrencyInitial.map((el) => {
      return {
        name: el.name,
        value: el.rate ? formatSum(el.rate * basicRate * time) : "--",
      };
    });
  }

  const basicRate = price / mainCurrencyData.rate;

  const mainCurrencySum = formatSum(time * price);

  const subCurrenciesArr = generateSubCurrenciesArr(allCurrencies, basicRate);

  const darkMode = context.darkMode;

  return (
    <div css={() => styles.getStyle(darkMode, "display")} className="display">
      {isRequestError && <p>{t("badRequestApi")}</p>}
      <MainCurrencyDisplay
        currency={mainCurrencyData.symbol}
        sum={mainCurrencySum}
        isLoading={isLoading}
        darkMode={darkMode}
      />
      <SubCurrenciesDisplay
        subCurrenciesArr={subCurrenciesArr}
        isLoading={isLoading}
        darkMode={darkMode}
      />
      <SubCurrenciesRatesDisplay
        allCurrencies={allCurrencies}
        currency={currency}
        darkMode={darkMode}
      />
    </div>
  );
}
