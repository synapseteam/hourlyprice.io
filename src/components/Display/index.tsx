/** @jsxImportSource @emotion/react */
import { FC } from "react";
import { useAppSelector } from "store/hooks";

import SubCurrenciesDisplay from "components/Display/SubCurrencies";
import MainCurrencyDisplay from "components/Display/MainCurrency/";
import SubCurrenciesRatesDisplay from "components/Display/SubCurrenciesRates";
import { DECIMAL_SIGNS_FOR_PRICE } from "utils/constants";

import { styles } from "./styles";
import { useTranslation } from "react-i18next";

const Display: FC = (): JSX.Element => {
  const [t] = useTranslation();
  const { price, time, currency } = useAppSelector(
    (state) => state.generic.fields
  );

  const allCurrencies = useAppSelector((state) => state.rates.allCurrencies);
  const isLoading = useAppSelector((state) => state.generic.isLoading);
  const isRequestError = useAppSelector(
    (state) => state.generic.ratesRequestErr
  );

  const mainCurrencyData = allCurrencies.find((el) => el.name === currency) || {
    name: "",
    rate: 0,
    symbol: "",
  };

  // TODO check any
  function formatSum(num: any) {
    return new Intl.NumberFormat("de-DE")
      .format(num.toFixed(DECIMAL_SIGNS_FOR_PRICE))
      .replace(".", " ");
  }

  // TODO check any
  function generateSubCurrenciesArr(arr: any[], basicRate: number) {
    const subCurrencyInitial = arr.filter((el) => el.name !== currency);
    if (!currency) {
      return subCurrencyInitial.map((el) => {
        return {
          ...el,
          value: 0,
        };
      });
    }

    return subCurrencyInitial.map((el) => {
      return {
        name: el.name,
        value: el.rate ? formatSum(el.rate * basicRate * Number(time)) : "--",
      };
    });
  }

  const basicRate = Number(price) / mainCurrencyData.rate;
  const mainCurrencySum = formatSum(Number(time) * Number(price));
  const subCurrenciesArr = generateSubCurrenciesArr(allCurrencies, basicRate);

  return (
    <div css={styles.display}>
      <MainCurrencyDisplay
        currency={mainCurrencyData.symbol}
        sum={mainCurrencySum}
        isLoading={isLoading}
      />
      <SubCurrenciesDisplay
        subCurrenciesArr={subCurrenciesArr}
        isLoading={isLoading}
      />
      {isRequestError && <p>{t("badRequestApi")}</p>}

      <SubCurrenciesRatesDisplay
        allCurrencies={allCurrencies}
        currency={currency}
      />
    </div>
  );
};

export default Display;
