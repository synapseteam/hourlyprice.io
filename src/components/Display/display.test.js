import React from "react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import renderer from "react-test-renderer";
import { AppThemeProvider } from "context/AppContext";

import Display from "components/Display";

if (!global.React) {
  global.React = React;
}

const mockStore = configureStore();
const initState = {
  main: {
    fields: {
      price: "",
      time: "",
      currency: "USD",
    },
  },
  rates: {
    allCurrencies: [
      { name: "USD", rate: 1, symbol: "$", isoCode: "840" },
      { name: "EUR", rate: 1, symbol: "€", isoCode: "978" },
      { name: "UAH", rate: 1, symbol: "₴", isoCode: "980" },
      { name: "RUB", rate: 1, symbol: "₽", isoCode: "810" },
    ],
    updatedAt: "",
    ratesSource: "MasterCard",
  },
};

const context = {
  darkMode: true,
};
const value = [context];

const store = mockStore(initState);

test("Display component snapshot test", () => {
  const tree = renderer
    .create(
      <AppThemeProvider value={value}>
        <Provider store={store}>
          <Display />
        </Provider>
      </AppThemeProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
