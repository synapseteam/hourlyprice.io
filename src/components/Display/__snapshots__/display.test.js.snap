// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display component snapshot test 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-1xt4dij-StyledDisplay{background-color:#31335a;border-radius:0.4rem;position:relative;width:100%;padding:3rem 2rem;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:column;-ms-flex-direction:column;flex-direction:column;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;-webkit-box-shadow:0px 18px 23px -3px rgba(0, 0, 0, 0.51);-moz-box-shadow:0px 18px 23px -3px rgba(0, 0, 0, 0.51);box-shadow:0px 18px 23px -3px rgba(0, 0, 0, 0.51);}",
      }
    }
    data-emotion="css 1xt4dij-StyledDisplay"
  />,
  <div
    className="display css-1xt4dij-StyledDisplay e1ppmzqf0"
  >
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-1y4cgmw-StyledTitle{margin:0;}",
        }
      }
      data-emotion="css 1y4cgmw-StyledTitle"
    />
    <h2
      className="css-1y4cgmw-StyledTitle eix4w6c2"
    >
      Total price
    </h2>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-1twvyoy-StyledContainer{border-bottom:1px solid lighten(#31335a, 15);min-height:6.8rem;}",
        }
      }
      data-emotion="css 1twvyoy-StyledContainer"
    />
    <div
      className="css-1twvyoy-StyledContainer eix4w6c1"
    >
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": ".css-9oghy1-StyledSum{color:#5599ec;font-weight:700;line-height:0.7;padding:2.4rem 1rem;margin:0;font-size:3rem;}",
          }
        }
        data-emotion="css 9oghy1-StyledSum"
      />
      <p
        className="css-9oghy1-StyledSum eix4w6c0"
      >
        0$
      </p>
    </div>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-1hjtp1r-StyledContainer{margin-top:1.6rem;margin-bottom:1rem;min-height:3rem;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;width:100%;-webkit-box-pack:space-around;-ms-flex-pack:space-around;-webkit-justify-content:space-around;justify-content:space-around;}",
        }
      }
      data-emotion="css 1hjtp1r-StyledContainer"
    />
    <div
      className="css-1hjtp1r-StyledContainer er5z01m0"
    >
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": ".css-tsn4jp-StyledSubCurrencyContainer{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:column;-ms-flex-direction:column;flex-direction:column;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;}",
          }
        }
        data-emotion="css tsn4jp-StyledSubCurrencyContainer"
      />
      <div
        className="css-tsn4jp-StyledSubCurrencyContainer eqbx3fg2"
      >
        <style
          dangerouslySetInnerHTML={
            Object {
              "__html": ".css-1nsjxjc-StyledSubCurrencyTitlte{color:#5599ec;font-size:0.8rem;}",
            }
          }
          data-emotion="css 1nsjxjc-StyledSubCurrencyTitlte"
        />
        <div
          className="css-1nsjxjc-StyledSubCurrencyTitlte eqbx3fg1"
        >
          EUR
        </div>
        <style
          dangerouslySetInnerHTML={
            Object {
              "__html": ".css-96u9oc-StyledSubCurrencySum{margin-top:0.4rem;font-weight:700;}",
            }
          }
          data-emotion="css 96u9oc-StyledSubCurrencySum"
        />
        <div
          className="css-96u9oc-StyledSubCurrencySum eqbx3fg0"
        >
          0
        </div>
      </div>
      <div
        className="css-tsn4jp-StyledSubCurrencyContainer eqbx3fg2"
      >
        <div
          className="css-1nsjxjc-StyledSubCurrencyTitlte eqbx3fg1"
        >
          UAH
        </div>
        <div
          className="css-96u9oc-StyledSubCurrencySum eqbx3fg0"
        >
          0
        </div>
      </div>
      <div
        className="css-tsn4jp-StyledSubCurrencyContainer eqbx3fg2"
      >
        <div
          className="css-1nsjxjc-StyledSubCurrencyTitlte eqbx3fg1"
        >
          RUB
        </div>
        <div
          className="css-96u9oc-StyledSubCurrencySum eqbx3fg0"
        >
          0
        </div>
      </div>
    </div>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-7235x7-StyledRatesContainer{background-color:#171d3d;width:100%;position:absolute;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-pack:center;-ms-flex-pack:center;-webkit-justify-content:center;justify-content:center;border-radius:0 0 0.4rem 0.4rem;bottom:0;}",
        }
      }
      data-emotion="css 7235x7-StyledRatesContainer"
    />
    <div
      className="css-7235x7-StyledRatesContainer e1orvfm72"
    >
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": ".css-1pmwpg8-StyledRateText{font-size:0.8rem;line-height:0.5;}",
          }
        }
        data-emotion="css 1pmwpg8-StyledRateText"
      />
      <p
        className="css-1pmwpg8-StyledRateText e1orvfm71"
      >
        €1.0000
        <style
          dangerouslySetInnerHTML={
            Object {
              "__html": ".css-16mfi3c-StyledRatesDelimiter{color:#5599ec;margin:0 0.6rem;}",
            }
          }
          data-emotion="css 16mfi3c-StyledRatesDelimiter"
        />
        <span
          className="css-16mfi3c-StyledRatesDelimiter e1orvfm70"
        >
          /
        </span>
      </p>
      <p
        className="css-1pmwpg8-StyledRateText e1orvfm71"
      >
        ₴1.0000
        <span
          className="css-16mfi3c-StyledRatesDelimiter e1orvfm70"
        >
          /
        </span>
      </p>
      <p
        className="css-1pmwpg8-StyledRateText e1orvfm71"
      >
        ₽1.0000
      </p>
    </div>
  </div>,
]
`;
